File: leetcode_ly/c++/DynamicProgramming/LargestDivisibleSubset.cpp
class Solution {
public:
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        int n = nums.size();
        vector<int> ans;
        if(n == 0) return ans;
        vector<int> res(n,1);
        vector<int> index(n,-1);
        quickSort(nums,0,nums.size() - 1);
        
        int max = 0; 
        int idx = 0;
        for(int i = 1; i < n; i++){
            for(int j = 0; j < i; j++){
                if(nums[i] % nums[j] == 0){
                    if(res[i] < res[j] + 1){
                        res[i] = res[j] + 1;
                        index[i] = j;
                    }
                }                
            }
            if(max < res[i]){
                max = res[i];
                idx = i;
            }
        }
        
        while(index[idx] != -1){
            ans.push_back(nums[idx]);
            idx = index[idx];
        }
        ans.push_back(nums[idx]);
        return ans;
    }
    
    void quickSort(vector<int>& nums,int left,int right){
        int low = left;
        int high = right;
        int target = nums[low];
        if(low < high){
            while(low < high){
                while(low < high && nums[high] >= target){
                    high--;
                }
                nums[low] = nums[high];
                while(low < high && nums[low] <= target){
                    low++;
                }
                nums[high] = nums[low];
            }
            nums[low] = target;
            quickSort(nums,left,low - 1);
            quickSort(nums,low + 1,right);
        }
    }
};

File: leetcode_ly/java/Backtracking && DP/LargestDivisibleSubset.java
public class Solution {
    public List<Integer> largestDivisibleSubset(int[] nums) {
        List<Integer> res = new ArrayList<Integer>();
        if(nums == null || nums.length == 0) return res;
        Arrays.sort(nums);
        int[] dp = new int[nums.length];
        Arrays.fill(dp,1);
        for(int i = 1; i < nums.length; i++){
            for(int j = 0; j < i; j++){
                if(nums[i] % nums[j] == 0) dp[i] = Math.max(dp[i],dp[j] + 1);
            }
        }
        
        int max = 0;
        int index = 0;
        for(int i = 0; i < dp.length; i++){
            if(dp[i] > max){
                max = dp[i];
                index = i;
            }
        }
        
        int count = dp[index];
        for(int i = index; i >= 0; i--){
            if(nums[index] % nums[i] == 0 && count == dp[i]){
                res.add(nums[i]);
                count--;
            }
        }
        return res;
    }
}